plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs

```diff
@@ -230,74 +193,82 @@ compileMarkedExprs opts markerName =
             _)
             -- value argument
             inner
-          | markerName == GHC.idName fid -> compileCoreExpr opts (show fs_locStr) codeTy inner
-      e@(GHC.Var fid) | markerName == GHC.idName fid -> failCompilationSDoc "Found invalid marker, not applied correctly" (GHC.ppr e)
-      GHC.App e a -> GHC.App <$> comp e <*> comp a
-      GHC.Lam b e -> GHC.Lam b <$> comp e
-      GHC.Let bnd e -> GHC.Let <$> compB bnd <*> comp e
+          | ?markerName == GHC.idName fid -> compileMarkedExpOrDefer (show fs_locStr) codeTy inner
+      e@(GHC.Var fid) | ?markerName == GHC.idName fid -> throwError . NoContext . InvalidMarkerError $ GHC.ppr e
+      GHC.App e a -> GHC.App <$> compileMarkers e <*> compileMarkers a
+      GHC.Lam b e -> GHC.Lam b <$> compileMarkers e
+      GHC.Let bnd e -> GHC.Let <$> compileBind bnd <*> compileMarkers e
       GHC.Case e b t alts -> do
-            e' <- comp e
-            let expAlt (a, bs, rhs) = (,,) a bs <$> comp rhs
+            e' <- compileMarkers e
+            let expAlt (a, bs, rhs) = (,,) a bs <$> compileMarkers rhs
             alts' <- mapM expAlt alts
             pure $ GHC.Case e' b t alts'
-      GHC.Cast e c -> flip GHC.Cast c <$> comp e
-      GHC.Tick t e -> GHC.Tick t <$> comp e
+      GHC.Cast e c -> flip GHC.Cast c <$> compileMarkers e
+      GHC.Tick t e -> GHC.Tick t <$> compileMarkers e
       e@(GHC.Coercion _) -> pure e
       e@(GHC.Lit _) -> pure e
       e@(GHC.Var _) -> pure e
       e@(GHC.Type _) -> pure e
 
--- Helper to avoid doing too much construction of Core ourselves
-mkCompiledCode :: forall a . BS.ByteString -> BS.ByteString -> CompiledCode PLC.DefaultUni a
-mkCompiledCode plcBS pirBS = SerializedCode plcBS (Just pirBS)
-
--- | Actually invokes the Core to PLC compiler to compile an expression into a PLC literal.
-compileCoreExpr :: (PluginOptions, GHC.FamInstEnvs) -> String -> GHC.Type -> GHC.CoreExpr -> GHC.CoreM GHC.CoreExpr
-compileCoreExpr (opts, famEnvs) locStr codeTy origE = do
+-- | Compile the core expression that is surrounded by a 'plc' marker,
+-- and return the plc compiled "bytecode" as a bytestring, to be injected back to the Haskell program.
+--
+-- If plugin-user has enabled 'defer-errors' and the PIR/PLC compilers have errored during typechecking/compilation,
+-- then the function returns a core expression that corresponds to a runtime error, to be injected back to the Haskell program.
+compileMarkedExpOrDefer :: (?opts :: PluginOptions, ?famEnvs :: GHC.FamInstEnvs)
+               => String -> GHC.Type -> GHC.CoreExpr -> ECore PLC.DefaultUni GHC.CoreExpr
+compileMarkedExpOrDefer locStr codeTy origE =
+    compileMarkedExpr locStr codeTy origE
+    & if poDeferErrors ?opts
+      then (`catchError` emitRuntimeError)
+      else id
```

Comment:
One reason I don't like this is that it makes the error handling look weirdly partial. In fact, any error is either used to fail compilation or put into the code, but here the two cases are separated.

Reply:
But the "meaning" of this is exactly the same as before this modification, correct?

Reply:
I think it probably does the right thing, but I had to think about it quite hard.

(Why do I care about how "obvious" it is that it does the right thing? a) It makes reading it easier in future, b) it makes it easier to change it without accidentally breaking things c) the more complicated the correctness reasoning is, the more likely you are to just make a mistake!)